tools/logger/filter.c

```diff
@@ -0,0 +1,172 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2020 Intel Corporation. All rights reserved.
+ *
+ * Author: Karol Trzcinski	<karolx.trzcinski@linux.intel.com>
+ */
+
+#include <smex/ldc.h>
+#include <sof/lib/uuid.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "convert.h"
+#include "filter.h"
+
+int parse_uuid(const char *src, struct sof_uuid *dst)
+{
+	uint32_t b;
+	uint32_t c;
+	uint32_t d0;
+	uint64_t d2;
+	int n;
+	int ret;
+
+	ret = sscanf(src, " <%X-%X-%X-%X-%lX>%n", &dst->a, &b, &c, &d0, &d2,
+		     &n);
+	if (ret != 5 || b > UINT32_MAX || c > UINT32_MAX)
+		return -EINVAL;
+	dst->b = b;
+	dst->c = c;
+	dst->d[0] = (d0 >> 8) & 0xFF;
+	dst->d[1] = d0 & 0xFF;
+	dst->d[2] = (d2 >> 40) & 0xFF;
+	dst->d[3] = (d2 >> 32) & 0xFF;
+	dst->d[4] = (d2 >> 24) & 0xFF;
+	dst->d[5] = (d2 >> 16) & 0xFF;
+	dst->d[6] = (d2 >> 8) & 0xFF;
+	dst->d[7] = d2 & 0xFF;
+	return n;
+}
+
+uint32_t translate_uuid(const struct snd_sof_uids_header *uids_dict,
+			const struct sof_uuid *uuid)
+{
+	struct sof_uuid_entry *entry;
+	int i;
+
+	entry = (struct sof_uuid_entry *)((uintptr_t)uids_dict +
+					 uids_dict->data_offset);
+
+	for (i = 0; i < uids_dict->data_length; i += sizeof(*entry)) {
+		if (memcmp(&entry->id, uuid, sizeof(*uuid)) == 0)
+			return get_uuid_key(uids_dict, entry);
+		++entry;
+	}
+	return 0;
+}
+
+/*
+ * Convert file with content formatted as:
+ *   "aaaaaaaa-bbbb-cccc-d0d1-d2d3d4d5d6d7 log_level ignored_content\n"
+ */
+int filter_update_firmware(const struct snd_sof_uids_header *uids_dict,
+			   const char *config_file)
+{
+	struct sof_uuid uuid;
+	uint32_t uuid_key;
+	FILE *config_fd;
+	ssize_t parsed;
+	int trace_lvl;
+	ssize_t read;
+	FILE *out_fd;
+	size_t len;
+	char *line;
+	int ret = 0;
+
+	config_fd = fopen(config_file, "r");
+	if (!config_fd) {
+		log_err(NULL, "unable to open %s file, %d", config_file, errno);
+		ret = -errno;
+		goto err;
+	}
+
+	out_fd = fopen(FILTER_KERNEL_PATH, "w");
+	if (!out_fd) {
+		log_err(NULL, "unable to open %s file, %d",
+			FILTER_KERNEL_PATH, errno);
+		ret = -errno;
+		goto err;
+	}
+
+	line = malloc(128);
+	if (!line) {
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	while (1) {
+		read = getline(&line, &len, config_fd);
+		if (read < 0)
+			break;
+		if (read < 2 || line[0] == '#')
+			continue;
+		parsed = parse_uuid(line, &uuid);
+		uuid_key = translate_uuid(uids_dict, &uuid);
+		if (parsed < 0 || !uuid_key) {
+			log_err(NULL, "Invalid uuid entry: %s", line);
```

Comment:
For compatibility I suggest moving the !uuid_key check lower, have it emit a warning instead of failing and do the fprintf on line 121 (as of this review) only if uuid_key != 0.